extensions/amp-audio/1.0/test/test-amp-audio.js

```diff
@@ -0,0 +1,385 @@
+      const audio = shadowRoot.querySelector('audio');
+
+      expect(audio.tagName).to.equal('AUDIO');
+      expect(element.getAttribute('width')).to.be.equal('503');
+      expect(element.getAttribute('height')).to.be.equal('53');
+      expect(audio.offsetWidth).to.be.greaterThan(1);
+      expect(audio.offsetHeight).to.be.greaterThan(1);
+      expect(audio.hasAttribute('controls')).to.be.true;
+      expect(audio.hasAttribute('autoplay')).to.be.true;
+      expect(audio.hasAttribute('muted')).to.be.true;
+      expect(audio.hasAttribute('preload')).to.be.true;
+      expect(audio.hasAttribute('loop')).to.be.true;
+      expect(audio.hasAttribute('src')).to.be.false;
+      expect(audio.childNodes[0].tagName).to.equal('SOURCE');
+      expect(audio.childNodes[0].getAttribute('src')).to.equal('audio.mp3');
+      expect(audio.childNodes[1].tagName).to.equal('SOURCE');
+      expect(audio.childNodes[1].getAttribute('src')).to.equal('audio.ogg');
+      expect(audio.childNodes[2].nodeType).to.equal(Node.TEXT_NODE);
+      expect(audio.childNodes[2].textContent).to.equal('Unsupported.');
+    });
+
+    it('should set its dimensions to the browser natural', async () => {
+      element = attachAndRun({
+        src: 'audio.mp3',
+      });
+
+      // Wait till rendering is finished
+      await waitForRender();
+
+      // Fetch audio element from shadowRoot
+      const {shadowRoot} = element;
+      const audio = shadowRoot.querySelector('audio');
+
+      expect(element.style.width).to.be.equal('300px');
+      expect(element.style.height).to.be.equal('30px');
+      if (/Safari|Firefox/.test(navigator.userAgent)) {
+        // Safari has default sizes for audio tags that cannot
+        // be overridden.
+        return;
+      }
+      expect(audio.offsetWidth).to.be.equal(300);
+      expect(audio.offsetHeight).to.be.equal(30);
+    });
+
+    it('should set its natural dimension only if not specified', async () => {
+      element = attachAndRun({
+        'width': '500',
+        src: 'audio.mp3',
+      });
+
+      // Wait till rendering is finished
+      await waitForRender();
+
+      expect(element.style.width).to.be.equal('500px');
+      expect(element.style.height).to.be.equal('30px');
+    });
+
+    it('should fallback when not available', async () => {
+      // For this single test, cause audio elements that are
+      // created to lack the necessary feature set, which should trigger
+      // fallback behavior.
+      const {createElement} = doc;
+      doc.createElement = (name) => {
+        if (name === 'audio') {
+          name = 'busted-audio';
+        }
+        return createElement.call(doc, name);
+      };
+
+      const element = doc.createElement('div');
+      element.toggleFallback = env.sandbox.spy();
+      const audio = new AmpAudio(element);
+      // audio.buildAudioElement();
+      await audio.layoutCallback();
+      expect(element.toggleFallback).to.be.calledOnce;
```

Comment:
It worked for me to attach the `toggleFallback` call to the audio's `onError` prop and then synthesize an `error` event in the test:

```suggestion
      element = attachAndRun({
        'width': '500',
        src: 'audio.mp3',
      });
      element.toggleFallback = env.sandbox.spy();
      await waitForRender();

      // Fetch audio element from shadowRoot
      const {shadowRoot} = element;
      const audio = shadowRoot.querySelector('audio');
      audio.dispatchEvent(new Event('error'));
      expect(element.toggleFallback).to.be.calledOnce;
```

This was in `component.js`:
```jsx
<audio ... onError={() => toggleFallback(true)}>
```